src/sentry/options/store.py

```diff
@@ -24,60 +26,149 @@
 
     def get(self, key):
+        """
+        Fetches a value from the options store.
+        """
+        result = self.get_cache(key)
+        if result is not None:
+            return result
+
+        result = self.get_store(key)
+        if result is not None:
+            return result
+
+        # As a last ditch effort, let's hope we have a key
+        # in local cache that's possibly stale
+        return self.get_local_cache(key, grace=True)
+
+    def get_cache(self, key):
+        """
+        First check agaist our local in-process cache, falling
+        back to the network cache.
+        """
+        value = self.get_local_cache(key)
+        if value is not None:
+            return value
+
+        cache_key = key.cache_key
         try:
-            result = self.get_cache(key)
+            value = self.cache.get(cache_key)
         except Exception:
             logger.warn(CACHE_FETCH_ERR, key.name, exc_info=True)
-            result = None
+            value = None
+        else:
+            if key.ttl > 0:
+                self._local_cache[cache_key] = _make_cache_value(key, value)
+        return value
+
+    def get_local_cache(self, key, grace=False):
+        """
+        Attempt to fetch a key out of the local cache.
+
+        If the key exists, but is beyond expiration, we only
+        return it if grace=True. This forces the key to be returned
+        in a disaster scenario as long as we're still holding onto it.
+        This allows the OptionStore to pave over potential network hiccups
+        by returning a stale value.
+        """
+        try:
+            value, expires, grace = self._local_cache[key.cache_key]
+        except KeyError:
+            return None
+
+        now = int(time())
 
-        if result is None:
+        # Key is within normal expiry window, so just return it
+        if now < expires:
+            return value
+
+        # If we're able to accept within grace window, return it
+        if grace and now < grace:
+            return value
+
+        # Let's clean up values if we're beyond grace.
+        if now > grace:
             try:
-                result = self.get_store(key)
-            except Option.DoesNotExist:
-                result = None
-            except Exception as e:
-                logger.exception(unicode(e))
-                result = None
-            else:
-                # we only attempt to populate the cache if we were previously
-                # able to successfully talk to the backend
-                # NOTE: There is definitely a race condition here between updating
-                # the store and the cache
-                try:
-                    self.set_cache(key, result)
-                except Exception:
-                    logger.warn(CACHE_UPDATE_ERR, key.name, exc_info=True)
-        return result
+                del self._local_cache[key.cache_key]
```

Comment:
Yeah, good point. I'll note it, but I'm trying not to introduce locking here since it's a "best effort". If the local cache got nuked by another thread by accident, the worst is it introduces a new network hop. Not terrible, so I don't think it's worth thread locals and locking imo.